dipy/reconst/ivim.py

```diff
@@ -509,6 +536,360 @@ def _leastsq(self, data, x0):
+
+        Returns
+        -------
+        f1, f2 (volume fractions)
+
+        Notes
+        --------
+        cost function for differential evolution algorithm:
+
+        .. math::
+
+            minimize(norm((signal)- (phi*f)))
+        """
+
+        # Create four scalar optimization variables.
+        f = cvxpy.Variable(2)
+        # Create four constraints.
+        constraints = [cvxpy.sum(f) == 1,
+                       f[0] >= 0.011,
+                       f[1] >= 0.011,
+                       f[0] <= 0.29,
+                       f[1] <= 0.89]
+
+        # Form objective.
+        obj = cvxpy.Minimize(cvxpy.sum(cvxpy.square(phi * f - signal)))
+
+        # Form and solve problem.
+        prob = cvxpy.Problem(obj, constraints)
+        prob.solve()  # Returns the optimal value.
+        return np.array(f.value)
+
+    def nlls_cost(self, x_f, signal):
+        """
+        Cost function for the least square problem. The cost function is used
+        in the Least Squares function of SciPy in :func: `fit`. It guarantees
+        that stopping point of the algorithm is at least a stationary point
+        with reduction in the the number of iterations required by the
+        differential evolution optimizer.
+
+        Parameters
+        ----------
+        x_f : array
+            Contains the parameters 'x' and 'f' combines in the same array.
+
+        signal : array
+            The signal values measured for this model.
+
+        Returns
+        -------
+        sum{(signal -  phi*f)^2}
+
+        Notes
+        --------
+        cost function for the least square problem.
+
+        .. math::
+
+            sum{(signal -  phi*f)^2}
+        """
+
+        x, f = self.x_f_to_x_and_f(x_f)
+        f1 = np.array([f, 1 - f])
+        phi = self.Phi(x)
+        return np.sum((np.dot(phi, f1) - signal) ** 2)
+
+    def x_f_to_x_and_f(self, x_f):
+        """
+        Splits the array of parameters in x_f to 'x' and 'f' for performing
+        a search on the both of them independently using the Trust Region
+        Method.
+
+        Parameters
+        ----------
+        x_f : array
+            Combined array of parameters 'x' and 'f' parameters.
+
+        Returns
+        -------
+        x, f : array
+            Splitted parameters into two separate arrays
+
+        """
+        x = np.zeros(2)
+        f = x_f[0]
+        x = x_f[1:3]
+        return x, f
+
+    def x_and_f_to_x_f(self, x, f):
+        """
+        Combines the array of parameters 'x' and 'f' into x_f for performing
+        NLLS on the final stage of optimization.
+
+        Parameters
+        ----------
+         x, f : array
+            Splitted parameters into two separate arrays
+
+        Returns
+        -------
+        x_f : array
+            Combined array of parameters 'x' and 'f' parameters.
+
+        """
+        x_f = np.zeros(3)
+        x_f[0] = f[0]
+        x_f[1:3] = x
+        return x_f
+
+    def Phi(self, x):
```

Comment:
Could we use lower case here? "phi" instead of "Phi"?

Reply:
Yes, will do!